### PR TITLE
Feature: "Impersonate user" setting on Datasource

### PR DESCRIPTION
**Pull request by dmigo**
Implements #2148 

* New setting for Datasources called "Impersonate user"
* Depending on the setting username is going to be included in the connection string for a **sync** execution
* Depending on the setting username is going to be included in the connection string for an **async** execution

All the code that makes difference is covered, I don't really think it would make sense to add any more tests as a part of the current PR.

### CODE REVIEW

**Comment by dmigo** [comment]
Current solution works for sync queries but doesn't for async. It says 

> '_AppCtxGlobals' object has no attribute 'user'

Which makes sense. My first idea would be to pass it as a parameter to `get_sql_results` for the async case.

**Comment by dmigo** [comment]
hmm... Tests seem to be kinda off. How could I restart them?

**Comment by dmigo** [comment]
@mistercrunch 
All the code that makes difference is covered, I don't really think it would make sense to add any more tests as a part of the current PR.
Travis fails with js tests, though I didn't touch any js in this PR.

**Comment by dmigo** [comment]
@mistercrunch
All the code that makes difference is covered, I don't really think it would make Sinn to add any more tests as a part of the current PR.

File: superset/models/core.py [hunk 1]
```diff
@@ -588,13 +589,15 @@ def set_sqlalchemy_uri(self, uri):
...
-    def get_sqla_engine(self, schema=None, nullpool=False):
+    def get_sqla_engine(self, schema=None, nullpool=False, user_name=None):
         extra = self.get_extra()
         uri = make_url(self.sqlalchemy_uri_decrypted)
         params = extra.get('engine_params', {})
         if nullpool:
             params['poolclass'] = NullPool
         uri = self.db_engine_spec.adjust_database_uri(uri, schema)
+        if self.impersonate_user:
+            uri.username = user_name if user_name != None else g.user.username
```
Comment (mistercrunch) [inline superset/models/core.py]: I prefer `user_name if user_name else g.user.username` as it accounts for empty strings and it's more pythonesque

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Minor change request, otherwise looks good.

Took me a moment to understand why you're passing the username around, and then I understood that in the async context the user isn't authenticated.

Good work!

File: superset/models/core.py (hunk 1 above)
Comment (dmigo) [reply superset/models/core.py]: done! 👍

**Comment by mjgp2** [comment]
Is there anything in the docs about this?

Seems like it sets the DB user in the connection string to the superset user name, but I don't understand how the password for that DB user is set?

**Pull request merged**